proximal/examples/test_deconv_sinv_psf.py

```diff
@@ -49,18 +54,22 @@ def adjoint(im_src, im_sim):
     
     return LinOpFactory(im_shape, im_shape, forward, adjoint)
 
-def cost_function(x, im, psf, mu=0.1): 
-    blur_op = my_factory(psf, im.shape)
-    im_blur = blur_op(x)
+def cost_function(x, im, psf, mu=0.1, using_px_conv=True): 
+    if using_px_conv:
+        im_blur = px.conv(psf2_cropped, x)
+    else:
+        blur_op = my_factory(psf, im.shape)
+        im_blur = blur_op(x)
+
     data_term = px.sum_squares(im_blur - im) 
-    im_grad = px.grad(im_blur)
-    grad_term = mu * px.norm1(im_grad) #+ (1 - mu) * px.sum_squares(im_grad)
+    im_grad = px.grad(x)
+    grad_term = mu * px.group_norm1(im_grad, group_dims=(2, )) #+ (1 - mu) * px.sum_squares(im_grad)
     return data_term + grad_term #+ px.nonneg(im_blur)
 
 # Solve
 x = px.Variable(im_sb.shape)
 prob = px.Problem(cost_function(x, im_sb, psf=psf2, mu=0.1))
-prob.solve(solver='pc', max_iters=2, verbose=True, x0=im_sb.copy())
+prob.solve(solver='pc', max_iters=20, verbose=True, x0=im_sb)
```

Comment:
Solver `PC` is inefficient. In order to fit the problem into `pc`, ProxImaL ended up implemented the updates with gradient descent algorithm, not proximal. Please try `admm` or `ladmm` solvers.

Here's the time log on my machine... Note the keyword `least_squares` in the log:

```
Estimated params [sigma = 1.000 | tau = 1.000 | theta = 1.000 | L_est = 1.0000]
op = pc_iteration_tot, evals = 20, total_time (ms) = 193882.8656640835, avg_time (ms) = 9694.143283204176
op = copyprev, evals = 20, total_time (ms) = 501.7119560507126, avg_time (ms) = 25.08559780253563
op = calcz, evals = 20, total_time (ms) = 6015.871784824412, avg_time (ms) = 300.7935892412206
op = calcx, evals = 20, total_time (ms) = 4501.566511928104, avg_time (ms) = 225.0783255964052
op = omega_fn, evals = 20, total_time (ms) = 140335.76565404655, avg_time (ms) = 7016.788282702328
op = xbar, evals = 20, total_time (ms) = 235.28147512115538, avg_time (ms) = 11.764073756057769
op = conv_check, evals = 1, total_time (ms) = 357.88110102294013, avg_time (ms) = 357.88110102294013
prox funcs total:
op = group_norm1, evals = 20, total_time (ms) = 41933.84755816078, avg_time (ms) = 2096.692377908039
op = least_squares, evals = 20, total_time (ms) = 140335.6384570361, avg_time (ms) = 7016.781922851806
prox funcs inner:
op = group_norm1, evals = 20, total_time (ms) = 38352.65889193397, avg_time (ms) = 1917.6329445966985
op = least_squares, evals = 20, total_time (ms) = 140334.86549992813, avg_time (ms) = 7016.743274996406
K forward ops:
op = vstack, evals = 22, total_time (ms) = 4935.982060793322, avg_time (ms) = 224.362820945151
op = scale, evals = 22, total_time (ms) = 292.875456972979, avg_time (ms) = 13.312520771499045
op = grad, evals = 22, total_time (ms) = 382.1921531925909, avg_time (ms) = 17.372370599663224
op = copy, evals = 22, total_time (ms) = 86.96805912768468, avg_time (ms) = 3.95309359671294
op = split, evals = 22, total_time (ms) = 804.5074680703692, avg_time (ms) = 36.568521275925875
op = CompGraph, evals = 22, total_time (ms) = 6505.099232075736, avg_time (ms) = 295.6863287307153
K adjoint ops:
op = vstack, evals = 20, total_time (ms) = 1944.3632349721156, avg_time (ms) = 97.21816174860578
op = scale, evals = 20, total_time (ms) = 253.46375803928822, avg_time (ms) = 12.673187901964411
op = grad, evals = 20, total_time (ms) = 380.49734686501324, avg_time (ms) = 19.024867343250662
op = copy, evals = 20, total_time (ms) = 72.6744559360668, avg_time (ms) = 3.6337227968033403
op = split, evals = 20, total_time (ms) = 1742.8693570545875, avg_time (ms) = 87.14346785272937
op = CompGraph, evals = 20, total_time (ms) = 4396.093876159284, avg_time (ms) = 219.8046938079642
```

Reply:
The deconvolution results after 20 iterations:

![image](https://github.com/comp-imaging/ProxImaL/assets/12185630/14dd5296-a99f-42ba-acd7-f66f87cd7f81)

Reply:
Thanks @antonysigma, I am not able to reproduce your results with the current commit to the branch. Can you please share the values of parameters you end up setting? Did you use `admm` and any other changes you made?

Reply:
Thanks @antonysigma, so I just played with your changes, and it turns out that the critical change was that of replacing `fftconvolve` with the built-in linear operator `convolve`. If I keep everything the same i.e. use `pc` as the algorithm and use anisotropic TV with `norm1()` instead of `group_norm1`, and just change the convolution operator then I get the results that you shared. 

So why isn't our implementation of the convolution linear operator by wrapping `forward` and `adjoint` in the `factory` function not working?

Reply:
> If I keep everything the same i.e. use pc as the algorithm and use anisotropic TV with norm1() instead of group_norm1, and just change the convolution operator then I get the expected results that you shared.

Great! To double confirm, did the solver reports a "slow path" in the update step? You should notice the  keyword `least_square` with a very large time spent per iteration (i.e. >7000ms per iter).

> So why isn't our implementation of the convolution linear operator by wrapping forward and adjoint in the factory function not working?

I am not sure. My computer is too slow; I didn't run the BlackBox to completion. What did you find on your end?

Reply:
@shnaqvi Update: perhaps the scaling/normalization step is skipped for `BlackBox` objects. How much do you know about **implicitly restarted Arnoldi method**? ProxImaL requires it to estimate the `norm_bound` of the `im_op` operation to guarantee convergence.

Or, manually approximate the scaling factor with `LinOpFactory(..., norm_bound=???)` to see the convergence improves.

References: https://github.com/comp-imaging/ProxImaL/blob/3437814f70b45669aab5a44f7da02ee2da5751db/proximal/lin_ops/black_box.py#L52-L68

Reply:
Well it runs pretty fast, like in 66sec, but yeah I do see `least_square` show up in the Time Log:
```
[Running] python -u "/ProxImaL/proximal/examples/test_deconv_sinv_psf.py"
Estimated params [sigma = 1.000 | tau = 1.000 | theta = 1.000 | L_est = 1.0000]
op = pc_iteration_tot, evals = 20, total_time (ms) = 54217.909957049415, avg_time (ms) = 2710.8954978524707
op = copyprev, evals = 20, total_time (ms) = 329.45258426479995, avg_time (ms) = 16.472629213239998
op = calcz, evals = 20, total_time (ms) = 2894.279956119135, avg_time (ms) = 144.71399780595675
op = calcx, evals = 20, total_time (ms) = 3210.3706264169887, avg_time (ms) = 160.51853132084943
op = omega_fn, evals = 20, total_time (ms) = 42329.22241440974, avg_time (ms) = 2116.461120720487
op = xbar, evals = 20, total_time (ms) = 91.91558591555804, avg_time (ms) = 4.595779295777902
op = conv_check, evals = 1, total_time (ms) = 209.10512504633516, avg_time (ms) = 209.10512504633516
prox funcs total:
op = group_norm1, evals = 20, total_time (ms) = 5153.092623804696, avg_time (ms) = 257.6546311902348
op = least_squares, evals = 20, total_time (ms) = 42329.139001783915, avg_time (ms) = 2116.4569500891957
prox funcs inner:
op = group_norm1, evals = 20, total_time (ms) = 5151.069169165567, avg_time (ms) = 257.5534584582783
op = least_squares, evals = 20, total_time (ms) = 42328.731375862844, avg_time (ms) = 2116.436568793142
K forward ops:
op = vstack, evals = 22, total_time (ms) = 186.3707520533353, avg_time (ms) = 8.47139782060615
op = scale, evals = 22, total_time (ms) = 138.53441714309156, avg_time (ms) = 6.297018961049616
op = grad, evals = 22, total_time (ms) = 2863.6462477734312, avg_time (ms) = 130.16573853515595
op = copy, evals = 22, total_time (ms) = 62.99308338202536, avg_time (ms) = 2.863321971910244
op = split, evals = 22, total_time (ms) = 63.621666864492, avg_time (ms) = 2.891893948386
op = CompGraph, evals = 22, total_time (ms) = 3316.1923331208527, avg_time (ms) = 150.73601514185694
K adjoint ops:
op = vstack, evals = 20, total_time (ms) = 72.18574685975909, avg_time (ms) = 3.6092873429879546
op = scale, evals = 20, total_time (ms) = 97.77941601350904, avg_time (ms) = 4.888970800675452
op = grad, evals = 20, total_time (ms) = 2899.951544124633, avg_time (ms) = 144.99757720623165
op = copy, evals = 20, total_time (ms) = 36.036292091012, avg_time (ms) = 1.8018146045506
op = split, evals = 20, total_time (ms) = 63.39120527263731, avg_time (ms) = 3.1695602636318654
op = CompGraph, evals = 20, total_time (ms) = 3170.17321055755, avg_time (ms) = 158.5086605278775

[Done] exited with code=0 in 66.574 seconds
```

Reply:
Sorry I misunderstood your question. Above time log is with using `px.conv()` with `pc`. 

Yes `BlackBox` based convolution operator definition is slow on my end too. With `eps` as the stopping criteria, it didn't stop in 10s of mins >60 items. So I forced `max_iters=20` and then it returned in 4mins.

Reply:
> perhaps the scaling/normalization step is skipped for BlackBox objects. How much do you know about implicitly restarted Arnoldi method? ProxImaL requires it to estimate the norm_bound of the im_op operation to guarantee convergence.

@antonysigma , thanks for that hypothesis. I'm not familiar with this _ implicitly restarted Arnoldi method_. I want to get the `BlackBox` approach to work because we have to use it to define our `linop` for the spatially variant PSF forward model.

Reply:
> I want to get the BlackBox approach to work because we have to use it to define our linop for the spatially variant PSF forward model.

Closing this thread regarding `least_square` and the slow path. 

Moving the `BlackBox` related question to https://github.com/comp-imaging/ProxImaL/pull/84#discussion_r1312377746